addons/account/models/res_partner_bank.py

```diff
@@ -253,6 +253,11 @@ def create(self, vals_list):
                 # force the allow_out_payment field to False in order to prevent scam payments on newly created bank accounts
                 vals['allow_out_payment'] = False
 
+        for vals in vals_list:
+            archived_res_partner_bank = self.env['res.partner.bank'].search([('active', '=', False), ('partner_id', '=', vals['partner_id']), ('acc_number', '=', vals['acc_number'])])
+            if archived_res_partner_bank:
+                raise UserError(_("A bank account with Account Number %(number)s already exists for Partner %(partner)s, but is archived. Please unarchive it instead.", number=vals['acc_number'], partner=vals['acc_holder_name']))
```

Comment:
Are you sure `vals['acc_holder_name']` is always set ? Looks like a non-required computed stored, so during creation could be not computed yet ?
Probably safer if you go with `archived_res_partner_bank.partner_id.name`.

I'd also add a similar check before your search: `if (partner_id := vals.get('partner_id')) and (acc_number := vals.get('acc_number')):` even when they are `required=True` to have a more robust code.